inc/az_ulib_descriptor_api.h

```diff
@@ -0,0 +1,210 @@
+
+#include "azure_macro_utils/macro_utils.h"
+#include "umock_c/umock_c_prod.h"
+
+#include "az_ulib_action_api.h"
+#include "az_ulib_base.h"
+#include "ulib_result.h"
+
+#ifndef __cplusplus
+#include <stddef.h>
+#include <stdint.h>
+#else
+#include <cstddef>
+#include <cstdint>
+extern "C" {
+#endif /* __cplusplus */
+
+/**
+ * @file    az_ulib_descriptor_api.h
+ *
+ * @brief   Contains the macros to creates the interface descriptor.
+ *
+ * This descriptor is used by the IPC to describe actions in the interface.
+ *
+ * The descriptor shall be created in compile time using this macros, which will place all the
+ * content of the descriptor in the Text area (as constant), avoiding copies and access violation.
+ */
+
+/**
+ * @brief   Action descriptor.
+ *
+ * This structure contains the full information about a single action in the interface. No matter
+ * the type of the action, it shall be fully described by this structure, which means that this
+ * structure is a superset of the fields of all types of actions.
+ */
+typedef struct az_ulib_action_descriptor_tag {
+  const char *name; /**<The `\0` terminated <tt>const char *</tt> with the action name.*/
+  const union /**<The primary function of the action. */
+  {
+    const void *action;
+    const az_ulib_action_get get;
+    const az_ulib_action_method method;
+    const az_ulib_action_method_async method_async;
+  } action_ptr_1;
+  const union /**<The secondary function of the action. */
+  {
+    const void *action;
+    const az_ulib_action_set set;
+    const az_ulib_action_cancelation_callback cancel;
+  } action_ptr_2;
+  const uint8_t flags; /**<This is a 8 bit flags that handles internal status of the action. */
+} az_ulib_action_descriptor;
+
+/**
+ * @brief   Interface descriptor.
+ *
+ * This structure contains the full information about a collection of action that represents the
+ * interface.
+ */
+typedef struct az_ulib_interface_descriptor_tag {
+  const char *name; /**<The `\0` terminated <tt>const char*</tt> with the interface name. */
+  az_ulib_version version; /**<The #az_ulib_version with the interface version. */
+  uint8_t size; /**<The <tt>uint8_t</tt> with the number of actions in the interface. */
+  az_ulib_action_descriptor *action_list; /**<The list of #az_ulib_action_descriptor with the
+                                          actions in this interface. */
+} az_ulib_interface_descriptor;
+
+/**
+ * @brief   Create a new Interface descriptor.
+ *
+ * This API creates a new description for a interface, it will create the
+ * #az_ulib_interface_descriptor with the provided interface name and version, adding all the
+ * actions.
+ *
+ * The descriptor will be stored in the Text area (Flash) that will be pointed by the provided
+ * interface_var. On this way, no other components on the system needs to copy any of the data on
+ * the descriptor.
+ *
+ * @param[in]   interface_var   The <tt>az_ulib_interface_descriptor</tt> the will pointe to the
```

Comment:
>the will pointe [](start = 74, length = 15)

that will point

Reply:
pointe

---
In reply to: [338204218](https://github.com/Azure/azure-ulib-c/pull/40#discussion_r338204218) [](ancestors = 338204218)